src/solvers/dgsem_tree/dg_2d.jl

```diff
@@ -172,7 +172,11 @@ function rhs!(du, u, t,
 
     # Calculate source terms
     @trixi_timeit timer() "source terms" begin
-        calc_sources!(du, u, t, source_terms, equations, dg, cache)
+        if size(cache.elements.node_coordinates, 1) == 3
+            calc_extended_sources!(du, u, t, source_terms, equations, dg, cache)
+        else
+            calc_sources!(du, u, t, source_terms, equations, dg, cache)
+        end
```

Comment:
Is there a better way to add a new source term function that takes in `du` as an input argument?

Context: In this implementation, we need a source term function that takes `du` as an input argument. This is used in this PR to transform the compressible Euler equations into the linear advection equations with variable advection velocity (used to test convergence). However, this is not limited to this application. In the future, a source term function of this form will allow an easy implementation of the Lagrange multiplier term for the SW equations on the sphere.

Reply:
Could you maybe elaborate a bit, please, so that I don't need to read the full diff of the PR? Right now, source terms are used to implement the $s$ in

$$\partial_t u(t, x) + \sum_j \partial_{x_j} f^j(u(t, x)) = s(t, x, u(t, x))$$

I don't see why we need `du` there. What's the form of equations you would like to discretize?

Reply:
There are two reasons why I want to use `du` for the source term:
1. For this particular PR, I wanted to test the convergence properties of the new implementation. However there is no 3D set of equations in Trixi.jl that can be used directly on the surface of the sphere to run a convergence test. I had two options: implement a new equation or use a source term that can transform an existing equation into an equation I need. The alternative that requires fewer lines of new code is to transform the 3D Euler equations into linear advection with *variable* advection velocity using a source term (note that I always need an advection velocity that is tangential to the sphere surface). To do that, I can simply use the following source term that needs `du`:
```julia
function source_terms_convert_to_linear_advection(u, du, x, t,
                                                  equations::CompressibleEulerEquations3D)
    v1 = u[2] / u[1]
    v2 = u[3] / u[1]
    v3 = u[4] / u[1]

    s2 = du[1] * v1 - du[2]
    s3 = du[1] * v2 - du[3]
    s4 = du[1] * v3 - du[4]
    s5 = 0.5 * (s2 * v1 + s3 * v2 + s4 * v3) - du[5]

    return SVector(0.0, s2, s3, s4, s5)
end
```
2. The simplest version of the SW equations with constant bottom topography on the sphere has the form 

$$ \partial_t u(t, x) + \nabla \cdot \vec{f} = s, $$ 

where the "source"/non-conservative term acts only in the momentum equations and has the form

$$ \vec{s}_{h\vec{v}} = \frac{\vec{x}}{\|\|\vec{x}\|\|^2} \vec{x} \cdot (\nabla \cdot \vec{\tilde{f}}), $$

where $\vec{\tilde{f}}$ represents the part of the flux associated with the momentum equations. The term $\vec{s}_{h\vec{v}}$ is a Lagrange multiplier term that ensures that the momentum stays tangential to the sphere surface. Note that this technique can also be used to constrain other equations (e.g., Euler) to the surface of the sphere.

The easiest and cheapest way to compute $\vec{s}_{h\vec{v}}$ is to use a source term function that takes `du` as an input argument.

Reply:
Could you please add this to the agenda for one of the next Trixi.jl meetings?
I can now see a reason why you want to use this form but I fear this might be a tool that has to be used with a lot of care.
If we allow this, we will probably have to accept to update the source terms everywhere (maybe losing performance a bit?) or maintain the increased complexity.

Reply:
I solved this using a custom RHS in a18fcd6. Thanks for the advice, @ranocha!